### PR TITLE
Createkeyrepo

### PR DESCRIPTION
**Pull request by joylatten**

### CODE REVIEW

File: createkeydb.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+		IsCA:			true,
+		BasicConstraintsValid: true,
+	}
+
+	rootCA, err := x509.CreateCertificate(rand.Reader, &caTemplate, &caTemplate, &privkey.PublicKey, privkey)
+	if err != nil {
+		return err
+	}
+	// Save the private key and cert to the specified directory
+	keyPEM, err := os.Create(filepath.Join(location, "key.pem"))
```
Comment (hallyn) [inline createkeydb.go:66]: What is going on here with tab/spacing?

File: utils.go [hunk 2]
```diff
@@ -49,15 +51,61 @@ func ConfPath(cluster string) string {
...
+	if keyrepo != "" && strings.HasPrefix(keyrepo, "/") {
+		if  PathExists(keyrepo) {
+			return keyrepo, nil
+		} else {
+			return "", fmt.Errorf("Specified keys repo %s does not exist.", keyrepo)
+		}
+	}
+
+	// Check if a github URL
+	if keyrepo != ""  && strings.HasPrefix(keyrepo, "https://github.com/") {
```
Comment (hallyn) [inline utils.go:107]: Probably shouldn't require github, right?

**Comment by hallyn** [comment]
Please use 'go fmt *.go' and also run golint.  While *I* personally prefer generate_rootca as a function name, golint will warn you to not use underscores.  They want camelCase.

File: createkeydb.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+    if err != nil {
+        return err
+    }
+    pkcs8, err := x509.MarshalPKCS8PrivateKey(privkey)
+    if err != nil {
```
Comment (hallyn) [inline createkeydb.go:71]: In this (and the next) case you are not closing keyPEM.

In theory golang will close it when it goes out of scope, but in practice (using a simple sleeping test) you can easily show that that doesn't happen very quickly.  The lack of 'defer keyPEM.Close()' at the top is in any case distracting so best to just have it there

File: utils.go (hunk 2 above)
Comment (joylatten) [reply utils.go:107]: ughhh i coded this on my other laptop... something up with my vimrc... let me check

File: createkeydb.go (hunk 3 above)
Comment (joylatten) [reply createkeydb.go:71]: totally missed that! yes, let me fix it.

File: keyset.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+	moskeysetPath, err := getMosKeyPath()
+	if err != nil {
+		return err
+	}
+	keysetPath := filepath.Join(moskeysetPath, keysetName)
+	if PathExists(keysetPath) {
+		return fmt.Errorf("%s keyset already exists", keysetName)
+	}
+
+	os.MkdirAll(keysetPath, 0755)
```
Comment (hallyn) [inline keyset.go:226]: Should this be 0750?

File: main.go [hunk 5]
```diff
@@ -168,168 +159,52 @@ func doInitrdSetup(ctx *cli.Context) error {
...
-	// Save the new key and signed certificate
 	trustDir, err := getTrustPath()
 	if err != nil {
 		return err
 	}
 
-	defer func() {
+	destdir := filepath.Join(trustDir, "manifest")
+	if ! PathExists(destdir) {
+		err = os.Mkdir(destdir, 0755)
```
Comment (hallyn) [inline main.go:183]: this also - should it be 0750?

**Comment by hallyn** [comment]
Looks fine apart from the two dir perms.

These really should all be squashed into one commit, right?

File: main.go (hunk 5 above)
Comment (joylatten) [reply main.go]: done

File: keyset.go (hunk 4 above)
Comment (joylatten) [reply keyset.go]: done

**Pull request merged**